### PR TITLE
Sidechannel modules (WIP)

### PR DESCRIPTION
**Pull request by mankenavenkatesh**

### CODE REVIEW

File: staking/side_handler.go [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+	// update last udpated
+	validator.LastUpdated = sequence.String()
+
+	// check if we are actually updating signer
+	if !bytes.Equal(newSigner.Bytes(), validator.Signer.Bytes()) {
+		// Update signer in prev Validator
+		validator.Signer = hmTypes.HeimdallAddress(newSigner)
+		validator.PubKey = newPubKey
+		k.Logger(ctx).Debug("Updating new signer", "signer", newSigner.String(), "oldSigner", oldValidator.Signer.String(), "validatorID", msg.ID)
+	}
```
[the remark this answers is not in the document]
Comment (jdkanani) [reply staking/side_handler.go:424]: @0xAshish we should also check if the current signer is not the same as new signer on contracts while updating signer changes.

cc @maticnetwork/contracts-devs

File: bridge/setu/listener/heimdall.go [hunk 2]
```diff
@@ -65,15 +65,34 @@ func (hl *HeimdallListener) StartPolling(ctx context.Context, pollInterval time.
...
+					if err != nil {
+						hl.Logger.Error("Error fetching begin block events", "error", err)
+					}
+					for _, event := range events {
+						hl.ProcessBlockEvent(sdk.StringifyEvent(event), int64(i))
+					}
+				}
+
+				// Querying and processing tx Events. Below for loop is kept for future purpose to process events from tx
 				for _, eventType := range eventTypes {
```
Comment (jdkanani) [inline bridge/setu/listener/heimdall.go:96]: Is this loop required?

File: bridge/setu/listener/heimdall.go [hunk 3]
```diff
@@ -65,15 +65,34 @@ func (hl *HeimdallListener) StartPolling(ctx context.Context, pollInterval time.
 	ticker := time.NewTicker(interval)
 
 	var eventTypes []string
-	eventTypes = append(eventTypes, "message.action='checkpoint'")
-	eventTypes = append(eventTypes, "message.action='event-record'")
+	// eventTypes = append(eventTypes, "message.action='checkpoint'")
+	// eventTypes = append(eventTypes, "message.action='event-record'")
+	// eventTypes = append(eventTypes, "message.action='tick'")
+	// ADD EVENT TYPE for SLASH-LIMIT
```
Comment (jdkanani) [inline bridge/setu/listener/heimdall.go:71]: Is this `eventTypes` required?

File: bridge/setu/listener/heimdall.go (hunk 2 above)
Comment (mankenavenkatesh) [reply bridge/setu/listener/heimdall.go:96]: It's kept for future incase we listen to tx events. 

currently it's of no use

File: topup/side_handler.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+			}
+		}
+	}
+}
+
+// NewPostTxHandler returns a side handler for "bank" type messages.
+func NewPostTxHandler(k Keeper, contractCaller helper.IContractCaller) hmTypes.PostTxHandler {
+	return func(ctx sdk.Context, msg sdk.Msg, sideTxResult abci.SideTxResultType) sdk.Result {
+		ctx = ctx.WithEventManager(sdk.NewEventManager())
+		fmt.Println("Topup PostTxHandler invoked", "msgType", msg.Type())
```
Comment (jdkanani) [inline topup/side_handler.go:40]: Remove unwanted fmt statements
```suggestion
```

File: staking/client/cli/tx.go [hunk 5]
```diff
@@ -304,13 +357,26 @@ func SendValidatorStakeUpdateTx(cdc *codec.Codec) *cobra.Command {
 	cmd.Flags().StringP(FlagProposerAddress, "p", "", "--proposer=<proposer-address>")
 	cmd.Flags().Int(FlagValidatorID, 0, "--id=<validator-id>")
 	cmd.Flags().String(FlagTxHash, "", "--tx-hash=<transaction-hash>")
+	cmd.Flags().String(FlagAmount, "", "--amount=<amount>")
 	cmd.Flags().String(FlagLogIndex, "", "--log-index=<log-index>")
+	cmd.Flags().Int(FlagBlockNumber, 0, "--block-number=<block-number>")
```
Comment (jdkanani) [inline staking/client/cli/tx.go:362]:
```suggestion
	cmd.Flags().UInt64(FlagBlockNumber, 0, "--block-number=<block-number>")
```

File: staking/client/cli/tx.go [hunk 6]
```diff
@@ -304,13 +357,26 @@ func SendValidatorStakeUpdateTx(cdc *codec.Codec) *cobra.Command {
 	cmd.Flags().StringP(FlagProposerAddress, "p", "", "--proposer=<proposer-address>")
 	cmd.Flags().Int(FlagValidatorID, 0, "--id=<validator-id>")
```
Comment (jdkanani) [inline staking/client/cli/tx.go:358]:
```suggestion
	cmd.Flags().UInt64(FlagValidatorID, 0, "--id=<validator-id>")
```

File: staking/client/cli/tx.go [hunk 7]
```diff
@@ -289,11 +334,19 @@ func SendValidatorStakeUpdateTx(cdc *codec.Codec) *cobra.Command {
...
+				return errors.New("Invalid new stake amount")
+			}
+
 			msg := types.NewMsgStakeUpdate(
 				proposer,
 				uint64(validator),
+				amount,
 				hmTypes.HexToHeimdallHash(txhash),
 				uint64(viper.GetInt64(FlagLogIndex)),
+				uint64(viper.GetInt64(FlagBlockNumber)),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:349]:
```suggestion
				viper.GetUInt64(FlagLogIndex),
				viper.GetUInt64(FlagBlockNumber),
```

File: staking/client/cli/tx.go [hunk 8]
```diff
@@ -252,6 +286,14 @@ func SendValidatorUpdateTx(cdc *codec.Codec) *cobra.Command {
 	cmd.Flags().String(FlagNewSignerPubkey, "", "--new-pubkey=<new-signer-pubkey>")
 	cmd.Flags().String(FlagTxHash, "", "--tx-hash=<transaction-hash>")
 	cmd.Flags().String(FlagLogIndex, "", "--log-index=<log-index>")
+	cmd.Flags().Int(FlagBlockNumber, 0, "--block-number=<block-number>")
```
Comment (jdkanani) [inline staking/client/cli/tx.go:289]:
```suggestion
	cmd.Flags().UInt64(FlagBlockNumber, 0, "--block-number=<block-number>")
```

File: staking/client/cli/tx.go [hunk 9]
```diff
@@ -240,6 +273,7 @@ func SendValidatorUpdateTx(cdc *codec.Codec) *cobra.Command {
 				pubkey,
 				hmTypes.HexToHeimdallHash(txhash),
 				uint64(viper.GetInt64(FlagLogIndex)),
+				uint64(viper.GetInt64(FlagBlockNumber)),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:276]:
```suggestion
				viper.GetUInt64(FlagLogIndex),
				viper.GetUInt64(FlagBlockNumber),
```

File: staking/client/cli/tx.go [hunk 10]
```diff
@@ -173,8 +198,10 @@ func SendValidatorExitTx(cdc *codec.Codec) *cobra.Command {
 			msg := types.NewMsgValidatorExit(
 				proposer,
 				uint64(validator),
+				uint64(viper.GetInt64(FlagDeactivationEpoch)),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:201]:
```suggestion
				viper.GetUInt64(FlagDeactivationEpoch),
```

File: staking/client/cli/tx.go [hunk 11]
```diff
@@ -173,8 +198,10 @@ func SendValidatorExitTx(cdc *codec.Codec) *cobra.Command {
...
 				hmTypes.HexToHeimdallHash(txhash),
 				uint64(viper.GetInt64(FlagLogIndex)),
+				uint64(viper.GetInt64(FlagBlockNumber)),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:204]:
```suggestion
				viper.GetUInt64(FlagBlockNumber),
```

File: staking/client/cli/tx.go [hunk 12]
```diff
@@ -136,6 +145,22 @@ func SendValidatorJoinTx(cdc *codec.Codec) *cobra.Command {
 	cmd.Flags().StringP(FlagProposerAddress, "p", "", "--proposer=<proposer-address>")
 	cmd.Flags().String(FlagSignerPubkey, "", "--signer-pubkey=<signer pubkey here>")
 	cmd.Flags().String(FlagTxHash, "", "--tx-hash=<transaction-hash>")
+	cmd.Flags().Int(FlagBlockNumber, 0, "--block-number=<block-number>")
```
Comment (jdkanani) [inline staking/client/cli/tx.go:148]:
```suggestion
	cmd.Flags().UInt64(FlagBlockNumber, 0, "--block-number=<block-number>")
```

File: staking/client/cli/tx.go [hunk 13]
```diff
@@ -173,8 +198,10 @@ func SendValidatorExitTx(cdc *codec.Codec) *cobra.Command {
 			msg := types.NewMsgValidatorExit(
 				proposer,
 				uint64(validator),
+				uint64(viper.GetInt64(FlagDeactivationEpoch)),
 				hmTypes.HexToHeimdallHash(txhash),
 				uint64(viper.GetInt64(FlagLogIndex)),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:203]:
```suggestion
				viper.GetUInt64(FlagLogIndex),
```

File: staking/client/cli/tx.go [hunk 14]
```diff
@@ -123,9 +129,12 @@ func SendValidatorJoinTx(cdc *codec.Codec) *cobra.Command {
 			msg := types.NewMsgValidatorJoin(
 				proposer,
 				event.ValidatorId.Uint64(),
+				viper.GetUint64(FlagActivationEpoch),
+				amount,
 				pubkey,
 				hmTypes.HexToHeimdallHash(txhash),
 				uint64(logIndex),
```
Comment (jdkanani) [inline staking/client/cli/tx.go:136]:
```suggestion
				logIndex,
```

File: staking/handler.go [hunk 15]
```diff
@@ -296,91 +140,13 @@ func HandleMsgSignerUpdate(ctx sdk.Context, msg types.MsgSignerUpdate, k Keeper,
 		k.Logger(ctx).Debug("Updating new signer", "signer", newSigner.String(), "oldSigner", oldValidator.Signer.String(), "validatorID", msg.ID)
 	}
```
[the remark this answers is not in the document]
Comment (mankenavenkatesh) [reply staking/handler.go:164]: Modified. 
 
https://github.com/maticnetwork/heimdall/blob/4a305f50e29e0f92258f809c0939fb0bb4183b8a/staking/handler.go#L162

File: staking/client/cli/tx.go (hunk 6 above)
Comment (mankenavenkatesh) [reply staking/client/cli/tx.go]: Resolved - https://github.com/maticnetwork/heimdall/blob/c567d4b05a59ea015c86e1f31e0b4fcebf12b67f/staking/client/cli/tx.go#L358

**Pull request merged**